### PR TITLE
fix: clarify tenant pool warming status and observability

### PR DESCRIPTION
**Pull request by srstack**
## Summary
- Return `creating` for admin tenant pools that already have reserved free slots but no active free tenant yet, so `pool create/get` no longer reports `active` with `free_size=0` during warmup.
- Keep the stored tenant-pool DB status as `active`; the `creating` value is response-only, so claim/replenish/delete state checks keep using the existing lifecycle.
- Stop tenant schema initialization when a pool resize/delete moves the tenant out of `provisioning`, avoiding repeated TiDB Cloud 1105 retries after shrink deletes a warming free tenant.
- Document and extend the tenant-pool lifecycle log surface: create/delete already emit `admin_tenant_pool_create_*` and `admin_tenant_pool_delete_*`; this PR adds `slot_size` to create completion and adds set-size `admin_tenant_pool_update_*` stage logs for grow/shrink/no-op/persist/fail paths.
- Record `drive9_service_operations_total` and `drive9_service_operation_duration_seconds` for `component="admin_tenant_pool"` across `operation="create|update|delete|replenish"`.
- Classify TiDB Cloud grow/shrink/delete/create failures as `result="cluster_error"` so lifecycle alerts can distinguish cloud/API failures from local metadata errors.

## Related Runbook Change
- Updated tenant-pool alerts in `tidbcloud/runbooks` branch `fix/drive9-worker-bad-conn-alert` / PR #2626.
- The failure alert now covers tenant pool `create|update|delete|replenish`, not just update/set-size.
- The latency alert remains focused on `operation="update"` because set-size holds the pool mutation lock around slow grow/shrink work.

## Tests
- `make lint`
- `make test TEST_PKGS='./pkg/server' TEST_RUN='TestTenantSchemaInitStopsWhenTenantLeavesProvisioning'`
- `make test TEST_PKGS='./pkg/server' TEST_RUN='TestAdminTenantPool|TestTenantSchemaInitStopsWhenTenantLeavesProvisioning'`
- YAML parse check for `/home/sweetqiu/project/runbooks/rules/mem9/mnemos/drive9-alerts.yaml`

## Notes
- `promtool` is not installed locally, so Prometheus rule validation was limited to YAML parsing.

### CODE REVIEW

**Review by srstack — COMMENTED** [review]
Reviewed all three commits. The `creating` display status and the lifecycle metrics/logs are a clean, well-scoped extension of the observability work — `adminTenantPoolDisplayStatus` is a pure response-only function (DB stays `active`, confirmed), the `active && freeSize==0 && slotSize>0` condition precisely captures "slots reserved, none claimable yet" = warming, and `adminTenantPoolMetricResult` correctly maps 502→`cluster_error`. Good. Three issues to fix before merge — the bots flagged the symptoms; here's the analysis and which fix is right.

### 1. Create silently degrades `creating`→`active` on a count error (CodeRabbit's fix is the correct one)
In create, `CountTenantPoolFreeSlots` errors are swallowed, leaving `slotSize=0`, so a freshly-warming pool is returned/logged as `active`. The GET path treats the same lookup as fatal (500), so **POST and GET now disagree for the same pool** — which is worse than the bug this PR fixes. CodeRabbit's suggestion `slotSize := len(results)` is the right fix and better than re-counting: at this point in create you *just created* `len(results)` slots, so it's the authoritative value and needs no DB round-trip at all. Use that instead of `CountTenantPoolFreeSlots` here.

### 2. `authorizedTenantPool` now couples authz with size-counting — regresses PATCH/DELETE (CodeRabbit, Major)
Adding the `slotSize` COUNT to `authorizedTenantPool` means **every** PATCH and DELETE now runs two COUNT queries it immediately discards (update re-counts inside the lock; delete ignores them entirely), and worse, introduces a **new pre-lock 500 path**: a transient `CountTenantPoolFreeSlots` failure now fails an unrelated DELETE that doesn't care about slot counts. Only GET needs the counts for display. Either return the counts only where used (a separate helper for GET), or have GET do its own counting after authorization. Keep authorization a pure authz check.

### 3. The end-to-end `creating` test is racy (Codex — confirmed)
`TestAdminTenantPool...` at quota_test.go:~1534 asserts `FreeSize==0 && Status==creating` after create. But create kicks off `startProvisionedTenantSchemaInit` async, and the fake provisioner's schema-init/system-user paths return immediately — so the goroutine can flip both tenants to `active` (free→2, status→active) before the GET runs. That makes the assertion schedule-dependent. The pure unit test `TestAdminTenantPoolDisplayStatusCreating` is fine (and good — keep it); it's this e2e one that needs the fake provisioner to hold tenants in pending/provisioning (block schema init / system-user setup) until after the GET. As written it'll flake in CI under load.

### Note on the metric defer (looks correct, calling it out)
The `defer metrics.RecordOperation(..., &metricResult, ...)` pattern with inner `errJSON` setting `metricResult` then `return nil` from the `WithTenantPoolLock` closure, plus the outer `if err != nil { metricResult = adminTenantPoolMetricResult(err) }`, doesn't double-count (single deferred record per handler) and resolves to the right label on both inner-rejection and lock/transaction-error paths. Confirmed fine.

Net: the feature is right and the observability is exactly what the companion runbooks alerts need. Fix #1 (`len(results)`) and #2 (decouple authz from counting) — both correctness/consistency, not polish — and de-flake the e2e test. The runbooks PR #2626 pairing looks appropriate (failure alert across create|update|delete|replenish, latency scoped to update since that's where the mutation lock wraps slow grow/shrink).

File: pkg/server/admin_tenant_pool.go [hunk 1]
```diff
@@ -205,25 +207,30 @@ func (s *Server) handleAdminTenantPoolCreate(w http.ResponseWriter, r *http.Requ
...
 		}
 		freeSize := 0
+		slotSize := 0
 		if orgID != "" {
 			if n, err := s.meta.CountFreeTenantPoolBindings(ctx, orgID); err == nil {
 				freeSize = n
 			}
+			if n, err := s.meta.CountTenantPoolFreeSlots(ctx, orgID); err == nil {
+				slotSize = n
+			}
```
[the remark this answers is not in the document]
Comment (srstack) [reply pkg/server/admin_tenant_pool.go:236]: Addressed in 1876870b: create now derives slot_size from len(results), so a slot-count lookup failure cannot degrade the response back to active. The free-size lookup now logs a warning if it fails.

**Review by srstack — COMMENTED** [review]
Re-reviewed after `1876870b` and `b7cdad29`. All three of my points are resolved cleanly, and `b7cdad29` is a good self-caught bug with coverage. No remaining blockers from me.

### Resolved
- **#1 create slot-count → `slotSize := len(results)`** (exactly the fix I recommended — authoritative, zero DB round-trip), and the free-size lookup error is now logged instead of swallowed. POST/GET status can no longer disagree from a silent count failure.
- **#2 authz/counting decoupled** — `authorizedTenantPool` is back to `(*meta.TenantPool, bool)` (pure authz), GET does its own counting via the new `tenantPoolSizes` helper, and update/delete no longer carry/recount discarded sizes. The pre-lock 500 path on unrelated DELETEs is gone.
- **#3 e2e de-flaked the right way** — `TestAdminTenantPoolGetShowsCreatingForPendingSlots` now directly seeds a pool + a `TenantPending` tenant with a `free` binding and hits GET, with no async create/schema-init goroutine in play. The `free_size=0, slot_size=1 → creating` assertion is now deterministic by construction, not by timing. Better than blocking the goroutine.
- Bonus: `Status` is now a typed `adminTenantPoolStatus` rather than a bare string — nice type-safety tightening across the response/display/delete paths.

### `b7cdad29` (stop schema init after shrink) — good catch, correct
This is a real bug the feedback work surfaced: if a free pool tenant is shrunk away (marked `deleting`) while its `initTenantSchemaAsync` is mid-flight, the worker would keep provisioning a tenant being torn down. The fix gates each stage (`before_attempt`, `before_finalize`, `after_error`) on `tenantSchemaInitStillProvisioning`, which re-reads tenant status and bails if it's no longer `TenantProvisioning` (or missing). Correctness notes I checked:
- It treats a transient `GetTenant` lookup error as "keep going" (returns true) rather than aborting — right call, avoids killing a valid init on a blip.
- `ErrNotFound` → stop, which is correct (tenant deleted out from under it).
- The regression test asserts `ensureSystemUserCalls == 0` after the tenant leaves `provisioning` — i.e. it verifies the *expensive* side effect is actually skipped, not just that status ends up `deleting`. Good test.
One thing worth confirming (not blocking): the status checks add a `GetTenant` read per stage per attempt. For a healthy provisioning tenant that's a few extra reads across the retry window — fine. Just make sure a persistently-failing schema init (many retries) doesn't multiply meta reads unboundedly; the retry backoff already bounds attempt count, so this is likely fine.

Net: feature + observability are solid, the three review points are addressed correctly, and the shrink-vs-schema-init race fix is a genuine improvement with the right test. LGTM from my side; pairs correctly with runbooks #2626. Confirm CI goes green (mid-run).

**Review by srstack — COMMENTED** [review]
Confirmed — all five match the code on the branch, and CI is now green (ci ✅, e2e smoke ✅, CodeRabbit ✅):

1. create response uses `slotSize := len(results)` — no `CountTenantPoolFreeSlots` dependency there.
2. `authorizedTenantPool` is pure authz again; GET counts via `tenantPoolSizes`; update/delete carry no unrelated counts.
3. racy post-create GET assertion replaced by `TestAdminTenantPoolGetShowsCreatingForPendingSlots` (direct-seeded pending free slot, no async goroutine) — deterministic by construction.
4. `adminTenantPoolStatus` is typed.
5. shrink/delete → `deleting` now stops schema init at `before_attempt`/`before_finalize`/`after_error` (via `tenantSchemaInitStillProvisioning`), and the test proves the expensive path is skipped (`ensureSystemUserCalls == 0`) — so no more wasted 1105 retries on a tearing-down tenant.

LGTM, no remaining blockers from me. Only the earlier non-blocking note stands (per-stage `GetTenant` adds a few meta reads per attempt, bounded by retry backoff — fine as-is). Pairs correctly with runbooks #2626.

**Review by qiffang — CHANGES_REQUESTED** [review]
## adversary-2 review — PR #646 (tenant pool warming status + observability)

**Verdict: REVISE** — 1 blocker

---

### B1 (blocker): `adminTenantPoolDisplayStatus` conflates "all claimed" with "still provisioning"

```go
func adminTenantPoolDisplayStatus(status meta.TenantPoolStatus, freeSize, slotSize int) adminTenantPoolStatus {
    if status == meta.TenantPoolActive && freeSize == 0 && slotSize > 0 {
        return adminTenantPoolStatusCreating
    }
    return adminTenantPoolStatus(status)
}
```

The condition `freeSize == 0 && slotSize > 0` is true in TWO distinct states:

1. **Warmup** — slots exist but tenants are still provisioning (intended case → `"creating"` is correct)
2. **Fully utilized** — all slots are claimed/used by workspaces, none are free (unintended case → `"creating"` is wrong; pool is actively serving)

`freeSize` = `CountFreeTenantPoolBindings` (bindings with pool_status = 'free')
`slotSize` = `CountTenantPoolFreeSlots` (total free-tier slot bindings regardless of pool_status)

When a pool of size 2 has all 2 tenants claimed, `freeSize = 0` and `slotSize` depends on whether claimed bindings are still counted as "free slots." If claimed bindings change pool_status away from `free`, then `slotSize` may also drop to 0 (masking the bug). But if `CountTenantPoolFreeSlots` counts all bindings regardless of pool_status, then `slotSize = 2, freeSize = 0` → the API reports `"creating"` for a fully utilized active pool.

The test suite (`TestAdminTenantPoolDisplayStatusCreating`) only tests the warmup case `(Active, 0, 2)` but not the fully-utilized case where all slots are claimed. This gap means the edge case is untested.

---

### Non-blockers

1. **`tenantSchemaInitStillProvisioning` — DB lookup on every retry attempt**: Each schema init retry calls `GetTenant` to check status. Under high-frequency retries (e.g., TiDB Cloud delays), this adds per-attempt DB load. Acceptable since retry intervals use exponential backoff (delays between attempts), but worth noting.

2. **`tenantPoolSizes` error messages strip original error**: `fmt.Errorf("tenant pool free size lookup failed")` discards the underlying error. The caller (`handleAdminTenantPoolGet`) passes this to `errJSON` so it's user-facing, but the original error is lost for debugging. Consider logging the original error before returning the sanitized message.

3. **Metrics instrumentation is comprehensive and correct**: Covers all exit paths (ok, conflict, error, cluster_error, skipped, noop, not_found). `adminTenantPoolMetricResult` correctly classifies BadGateway as `cluster_error` vs general errors.

---

### What's good

- Response-only status change (DB status remains `active`) preserves existing lifecycle state checks
- `tenantSchemaInitStillProvisioning` correctly handles: context cancelled, tenant not found, status changed, and DB lookup failure (fail-open to avoid blocking init on transient errors)
- Tests cover: display status transitions, GET with pending slots, schema init cancellation on status change, and existing tests updated to verify stored DB status vs response status divergence
- `authorizedTenantPool` refactor cleanly separates auth from size queries

**Review by qiffang — APPROVED** [review]
## dat9-dev1 — GREEN

**Verdict: GREEN — no blockers**

3 files changed, 432 additions, 36 deletions. Clean, well-scoped PR with three orthogonal changes.

### What it does

1. **Response-only `creating` status**: `adminTenantPoolDisplayStatus(active, freeSize=0, slotSize>0)` → `creating`. DB status remains `active`; claim/replenish/delete lifecycle checks are unaffected. Correctly distinguishes "slots reserved but still warming" from "pool ready with free tenants".

2. **Schema init guard**: `tenantSchemaInitStillProvisioning` checks tenant status at three points in the retry loop (before_attempt, before_finalize, after_error). When a pool shrink/delete moves the tenant out of `provisioning`, the init loop exits instead of hammering TiDB Cloud with 1105 retries. On DB lookup failure (non-ErrNotFound), returns `true` — safe default that preserves retry behavior.

3. **Lifecycle metrics + logging**: `RecordOperation` with `component="admin_tenant_pool"` across `create|update|delete|replenish`. `adminTenantPoolMetricResult` classifies `StatusBadGateway` as `cluster_error` to distinguish cloud API failures from local metadata errors. Structured stage logs (grow_started/done, shrink_started/done, noop, persist) cover the full update path.

### Design notes

- `authorizedTenantPool` refactored to return `(*meta.TenantPool, bool)` — size counting moved to separate `tenantPoolSizes` helper. Clean separation.
- Update handler now refreshes both `freeSize` and `slotSize` from DB **after** `UpdateTenantPoolSize` persists, ensuring the response reflects post-mutation state.
- Create handler uses `slotSize := len(results)` (batch provision count) instead of DB query — correct since these tenants were just created.
- `deleteNewestFreePoolTenants` return value now captured for `deleted_count` logging in shrink_failed — useful for partial failure diagnosis.

### Tests

- `TestAdminTenantPoolDisplayStatusCreating`: 4 cases covering the status matrix (pending slots, free slots, empty, deleting)
- `TestAdminTenantPoolGetShowsCreatingForPendingSlots`: End-to-end GET with pending tenant → `creating`
- `TestTenantSchemaInitStopsWhenTenantLeavesProvisioning`: Schema init exits when tenant status changes mid-retry
- Existing create/metadata tests updated to assert `creating` status and DB `active` storage
- Metrics assertion in `TestAdminTenantPoolUpdateShrinkPartialFailureReconcilesSize`

### Nits

1. `tenantPoolSizes` wraps DB errors into `fmt.Errorf("tenant pool free size lookup failed")` — loses the original error. Consider `fmt.Errorf("tenant pool free size lookup failed: %w", err)` so callers/logs can distinguish transient DB errors from permanent ones.

Head: `b7cdad299b13a121a87a275a00a4959273224fb2`

**Review by qiffang — APPROVED** [review]
## adversary-2 revised review — PR #646

**Verdict: GREEN** (revised from REVISE)

Previous B1 (display status conflation) withdrawn: `CountTenantPoolFreeSlots` filters on `pool_status = 'free'`, so fully-utilized pools (all bindings `pool_status = 'used'`) have `slotSize = 0` → `slotSize > 0` is false → correctly returns `active`. The `creating` status only fires during warmup when free-pool bindings exist but tenants are still `pending`/`provisioning`.

No blockers remain.

**Review by qiffang — APPROVED** [review]
## adversary-1 review — PR #646

Verdict: GREEN — no blockers.

Checked latest head `b7cdad299b13a121a87a275a00a4959273224fb2`.

What I verified:
- The `creating` status is response-only. Stored tenant-pool status remains `active`, so existing claim/replenish/delete lifecycle checks are not changed.
- The warmup predicate is valid: `adminTenantPoolDisplayStatus(active, freeSize=0, slotSize>0)` only fires when there are free pool bindings whose tenants are not yet claimable. `CountTenantPoolFreeSlots` delegates to `countFreeTenantPoolBindingsByStatus`, and that SQL filters `b.pool_status = TenantPoolBindingFree`; `ClaimOldestFreeTenantPoolBinding` moves claimed bindings to `TenantPoolBindingUsed`. So a fully utilized pool has `slotSize=0`, not `slotSize>0`, and remains displayed as `active`.
- Auth and size counting are decoupled. `authorizedTenantPool` is back to pure authorization/pool lookup; only GET uses `tenantPoolSizes`, and update recomputes sizes after mutation under the pool lock.
- The schema-init guard is directionally correct: before attempt, before finalize, and after error it stops when the tenant leaves `provisioning`, while transient lookup errors fail open to preserve existing retry behavior.
- Regression tests cover the display helper matrix, deterministic GET for pending free slots, create/metadata stored-`active` invariants, shrink partial failure metrics, and schema-init stop when status changes.

Validation:
- `git diff --check 8696e53235fe1344d3bf90fce2e70e14a9303062...HEAD` passed.
- GitHub checks are green: `ci`, `drive9-server-local e2e smoke`, CodeRabbit.
- I could not run local Go tests on this machine because `go` is not installed.

Non-blocking note: `tenantPoolSizes` returns sanitized lookup errors without wrapping/logging the original error. That is acceptable for the user-facing response, but if this path starts failing in production, adding server-side logging or `%w` wrapping would improve diagnostics.
